std/parallelism.d

```diff
@@ -2803,6 +2812,120 @@ public:
+                          format("Invalid number of arguments (%s): Should be an input range, %s optional seed(s)," ~
+                                 " and an optional work unit size", Args.length, functions.length));
+            return fold(args[0], args[1..$]);
+        }
+
+        /** This is the overload that uses implicit seeds. (See important notes
+            on implicit seeds under $(LREF reduce).)
+
+            Params:
+                range = The range of values to _fold over
+        */
+        auto fold(R)(R range)
+        {
+            return reduce!functions(range);
+        }
+
+        ///
+        unittest
+        {
+            auto result = taskPool.fold!"a+b"([1, 2, 3, 4]);
+            assert(result == 10);
+        }
+
+        /** This is the overload that uses explicit _seeds. (See important notes
+            on explicit _seeds under $(LREF reduce).)
+
+            Params:
+                range = The range of values to _fold over
+                seeds = One seed per function
+        */
+        auto fold(R, Seeds...)(R range, Seeds seeds)
+        if (Seeds.length == functions.length)
+        {
+            static if (Seeds.length == 1)
+            {
+                return reduce!functions(seeds, range);
+            }
+            else
+            {
+                import std.typecons : tuple;
+                return reduce!functions(tuple(seeds), range);
+            }
+        }
+
+        ///
+        unittest
+        {
+            auto result = taskPool.fold!("a+b", "a*b")([1, 2, 3, 4], 0, 1);
+            assert(result[0] == 10);
+            assert(result[1] == 24);
+        }
+
+        /** This is the overload that uses explicit seeds and work unit size.
+            (See important notes on explicit seeds and work unit size under
+            $(LREF reduce).)
+
+            Params:
+                range = The range of values to _fold over
+                args = One seed per function ($(D args[0..$-1])) and the work unit size ($(D args[$-1]))
+        */
+        auto fold(R, Args...)(R range, Args args)
+        if (Args.length == functions.length + 1)
+        {
+            static assert(isIntegral!(Args[$-1]), "Work unit size must be an integral type");
+
+            static if (Args.length == 2)
+            {
+                return reduce!functions(args[0], range, args[1]);
+            }
+            else
+            {
+                import std.typecons : tuple;
+                return reduce!functions(tuple(args[0..$-1]), range, args[$-1]);
+            }
+        }
+
+        ///
+        unittest
+        {
+                auto x = taskPool.fold!"a+b"([1, 2, 3, 4], 0, 20);                // Single function
+                assert(x == 10);
+
+                auto y = taskPool.fold!("a+b", "a*b")([1, 2, 3, 4], 0, 1, 30);    // Multiple functions
+                assert(y[0] == 10);
+                assert(y[1] == 24);
+        }
+    }
+
```

Comment:
Thanks. And making sure the unittests run exposed the infamous 

"Error: template instance fold!(function (a, b) => a + b) cannot use local '__funcliteral1' as parameter to non-global template fold(functions...)"

which means, I have to pull all unittests out to the module scope, which means, I can't take advantage of automatically inserting unittests as examples right under what they test. I think I will have to copy the code under Examples sections as had been done for TaskPool.reduce().

Any ideas for a solution?

Reply:
The usual practice is to put the unittest block(s) right after the template declaration. This way the examples will still get inserted into the docs for the template. Which is fine in this case, because it's not a big deal to just leave the inner template with a blank `///`, since the bulk of the docs will be on the outer template.

Furthermore, putting unittests inside a template is generally not a wise idea in terms of code bloat / possibly compilation time, because they will get instantiated for *every* instantiation of the template. Somebody tried to propose a `static unittest` sometime ago but it was shot down by Walter. Maybe we could try changing Walter's opinion on this again? :-P  But in the meantime, keeping unittests in module scope seems to be the best compromise.

Reply:
P.S. There's the workaround of using static if on a unittest inside a template so that it only gets instantiated for a specific template instantiation (i.e., only once), but that in turn requires something outside the template to trigger that specific instantiation in the first place, which makes it rather fragile.

Reply:
> Furthermore, putting unittests inside a template is generally not a wise idea in terms of code bloat / possibly compilation time, because they will get instantiated for every instantiation of the template. 

FYI @quickfur we have `StdUnittest` for this purpose now (see https://github.com/dlang/phobos/pull/5927).